File: day12/part1/main.dart
// -*- compile-command: "dart main.dart ../example1.txt ../input.txt"; -*-

import 'dart:io';

main(List<String> args) {
  for (var arg in args) {
    process(arg);
  }

  print('Done.');
}

process(String filename) {
  print('Processing $filename ...');

  var puz = Puzzle(filename);

  puz.iterate();

  var dist = puz.manhattan();

  print('Solution: $dist');
}

class Puzzle {
  int de = 0;
  int dn = 0;
  int epos = 0;
  int npos = 0;
  List<String> lines = [];

  Puzzle(String filename) {
    lines = File(filename).readAsLinesSync();
    de = 1;
  }

  iterate() {
    for (var line in lines) {
      var amt = int.parse(line.substring(1));
      switch (line.substring(0, 1)) {
        case 'N':
          npos += amt;
          break;
        case 'S':
          npos -= amt;
          break;
        case 'E':
          epos += amt;
          break;
        case 'W':
          epos -= amt;
          break;
        case 'L':
          while (amt > 0) {
            var tmp = dn;
            dn = de;
            de = -tmp;
            amt -= 90;
          }
          break;
        case 'R':
          while (amt > 0) {
            var tmp = dn;
            dn = -de;
            de = tmp;
            amt -= 90;
          }
          break;
        case 'F':
          npos += amt * dn;
          epos += amt * de;
          break;
      }
      // print('$line: ($epos,$npos) ($de,$dn)');
    }
  }

  int manhattan() {
    var x = epos;
    if (x < 0) {
      x = -x;
    }
    var y = npos;
    if (y < 0) {
      y = -y;
    }
    return x + y;
  }
}

File: day12/part1/main.go
// -*- compile-command: "go run main.go ../example1.txt ../input.txt"; -*-

package main

import (
	"flag"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
)

var (
	verbose = flag.Bool("v", false, "Verbose log messages")
)

func main() {
	flag.Parse()

	for _, arg := range flag.Args() {
		process(arg)
	}

	log.Printf("Done.")
}

func process(filename string) {
	puz := readPuzzle(filename)

	puz.iterate()

	dist := puz.manhattan()

	log.Printf("Solution: %v", dist)
}

func (p *Puzzle) manhattan() int {
	x := p.epos
	if x < 0 {
		x = -x
	}
	y := p.npos
	if y < 0 {
		y = -y
	}
	return x + y
}

func (p *Puzzle) iterate() {
	for _, line := range p.lines {
		if line == "" {
			continue
		}

		amt, err := strconv.Atoi(line[1:])
		check("Atoi: %v", err)

		switch line[0:1] {
		case "N":
			p.npos += amt
		case "S":
			p.npos -= amt
		case "E":
			p.epos += amt
		case "W":
			p.epos -= amt
		case "L":
			for amt > 0 {
				p.de, p.dn = -p.dn, p.de
				amt -= 90
			}
		case "R":
			for amt > 0 {
				p.de, p.dn = p.dn, -p.de
				amt -= 90
			}
		case "F":
			p.npos += amt * p.dn
			p.epos += amt * p.de
		}
		// log.Printf("%v: (%v,%v) (%v,%v)", line, p.epos, p.npos, p.de, p.dn)
	}
}

type Puzzle struct {
	de    int
	dn    int
	epos  int
	npos  int
	lines []string
}

func readPuzzle(filename string) *Puzzle {
	logf("Processing %v ...", filename)
	buf, err := ioutil.ReadFile(filename)
	check("ReadFile: %v", err)

	lines := strings.Split(string(buf), "\n")
	puz := &Puzzle{lines: lines, de: 1}

	return puz
}

func check(fmtStr string, args ...interface{}) {
	if err := args[len(args)-1]; err != nil {
		log.Fatalf(fmtStr, args...)
	}
}

func logf(fmt string, args ...interface{}) {
	if *verbose {
		log.Printf(fmt, args...)
	}
}
